usage_example/devbuf_map.c

```diff
@@ -0,0 +1,238 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <stdint.h>
+#include <stdarg.h>
+#include <string.h>
+#include <getopt.h>
+#include <limits.h>
+#include <assert.h>
+#include <sys/types.h>
+#include <sys/errno.h>
+#include <sys/fcntl.h>
+#include <sys/mman.h>
+#include <inttypes.h>
+#include <string.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <sys/ioctl.h>
+#include <linux/types.h>
+#include <linux/ioctl.h>
+
+/*
+ * Definitions for using IOCTL commands to get the buffer details
+ */
+#define U_DMA_BUF_IOCTL_MAGIC               'U'
+#define U_DMA_BUF_IOCTL_GET_SIZE            _IOR (U_DMA_BUF_IOCTL_MAGIC, 2, uint64_t)
+#define U_DMA_BUF_IOCTL_GET_DMA_ADDR        _IOR (U_DMA_BUF_IOCTL_MAGIC, 3, uint64_t)
+
+/*
+ * Structure to hold the device buffer details.
+ * Only the start physical address and the size of the buffer is used for now.
+ */
+
+typedef struct devbuf_details_t{
+	uint64_t  start_phys_addr;
+	size_t    size;
+}devbuf_details;
+
+/*
+ * Open a file and check for errors.
+ * 
+ * @param pathname: Path to the file to open
+ * @param flags: Flags to open the file with
+ * 
+ * @return: File descriptor on success 
+ */
+static int checked_open(const char *pathname, int flags)
+{
+	int fd = open(pathname, flags);
+
+	if (fd < 0) {
+		perror(pathname);
+		exit(EXIT_FAILURE);
+	}
+
+	return fd;
+}
+
+/*  
+ * Get the buffer details for a device buffer.
+ * Uses IOCTL commands to get the buffer size and the physical address of the buffer.
+ * 
+ * @param fd: File descriptor of the device buffer
+ * @param devbuf: Pointer to the devbuf_details structure to store the buffer details
+ * 
+ * @return: 0 on success, -1 on failure
+ */
+
+static int get_device_buffer_details_filedes(int fd, devbuf_details *devbuf)
+{
+    int ret;
+    uint64_t size;
+    uint64_t phys_addr;
+
+    ret = ioctl(fd, U_DMA_BUF_IOCTL_GET_SIZE, &size);
+    if (ret < 0) {
+        perror("ioctl U_DMA_BUF_IOCTL_GET_SIZE");
+        return -1;
+    }
+    printf("buf_size: %lu\n", size);
+
+    ret = ioctl(fd, U_DMA_BUF_IOCTL_GET_DMA_ADDR, &phys_addr);
+    if (ret < 0) {
+        perror("ioctl U_DMA_BUF_IOCTL_GET_DMA_ADDR");
+        return -1;
+    }
+    printf("phys_addr: %lx\n", phys_addr);
+
+    if (devbuf)
+    {
+        devbuf->size = size;
+		devbuf->start_phys_addr = phys_addr;
+    }
+    return 0;
+}
+
+/*
+* Get the buffer details for a device buffer.
+* 
+* @param buf_num: Number of the device buffer
+* @param devbuf: Pointer to the devbuf_details structure to store the buffer details
+* 
+* @return: 0 on success, -1 on failure    
+
+*/
+static int get_device_buffer_details(int buf_num, devbuf_details *devbuf)
+{
+    char dev_path[BUFSIZ];
+    int ret;
+    uint64_t size;
+    uint64_t phys_addr;
+    int fd;
+
+    snprintf(dev_path, sizeof(dev_path), "/dev/udmabuf%d", buf_num);
+    fd = checked_open(dev_path, O_RDWR);
+
+	if (fd != -1) {
+        ret = get_device_buffer_details_filedes(fd, devbuf);
```

Comment:
ret is set, but never used. You probably need to do "return ret;" at the end?

Reply:
Will change this to return the value of ret.